ols/src/query_helpers/query_docs.py

```diff
@@ -0,0 +1,124 @@
+        search_type: str = "similarity",
+        **search_kwargs: Any,
+    ) -> List[Document] | None:
+        """Return list of documents from the vectorstore.
+
+        Args:
+            query: string to find relevant documents for
+            perform_query: Whether RAG retrieval needs to be done or not.
+            vectordb: Vector db from which you need to retrieve the docs.
+            search_type (Optional[str]): Defines the type of search that
+                the retriever should perform.
+                Can be "similarity" (default), "mmr", or
+                "similarity_score_threshold".
+            search_kwargs (Optional[Dict]): Keyword arguments to pass to the
+                search function. Can include things like:
+                    k: Amount of documents to return (Default: 4)
+                    score_threshold: Minimum relevance threshold
+                        for similarity_score_threshold
+                    fetch_k: Amount of documents to pass to MMR algorithm (
+                    Default: 20)
+                    lambda_mult: Diversity of results returned by MMR;
+                        1 for minimum diversity and 0 for maximum. (Default:
+                        0.5)
+                    filter: Filter by document metadata
+
+        Returns:
+            List[Document]: Returns a list of docs.
+
+        Examples:
+        .. code-block:: python
+
+            # Retrieve more documents with higher diversity
+            # Useful if your dataset has many similar documents
+            docsearch.as_retriever(
+                search_type="mmr",
+                search_kwargs={'k': 6, 'lambda_mult': 0.25}
+            )
+
+            # Fetch more documents for the MMR algorithm to consider
+            # But only return the top 5
+            docsearch.as_retriever(
+                search_type="mmr",
+                search_kwargs={'k': 5, 'fetch_k': 50}
+            )
+
+            # Only retrieve documents that have a relevance score
+            # Above a certain threshold
+            docsearch.as_retriever(
+                search_type="similarity_score_threshold",
+                search_kwargs={'score_threshold': 0.8}
+            )
+
+            # Only get the single most similar document from the dataset
+            docsearch.as_retriever(search_kwargs={'k': 1})
+
+            # Use a filter to only retrieve documents from a specific paper
+            docsearch.as_retriever(
+                search_kwargs={'filter': {'paper_title':'GPT-4 Technical
+                Report'}}
+            )
+        """
+        self.logger.info(
+            f"""Retrieving docs for
+                query: {query},
+                perform_query: {perform_query},
+                vectordb: {vectordb},
+                search_type: {search_type},
+                search_kwargs: {search_kwargs}
+        """
+        )
+
+        if not perform_query:
+            return None
```

Comment:
This is one of the acceptance criteria. This was specifically discussed in backlog refinement and was asked to be part of this component

Reply:
+1 @onmete 

retrieval should be optional means "in the overall OLS flow, when a user sends us a question there may be cases where we cannot do, or choose not to do, retrieval augmented generation for that particular query" (maybe because no RAG index is configured, maybe because we understand the question won't benefit from it, maybe some other reason).

But that decision would be made closer to the request flow and the logic would simply not invoke `get_relevant_docs`.  Having logic that calls `get_relevant_docs` while passing an argument that says "but don't actually retrieve anything" doesn't make much sense....  you should just not make the function call in the first place.

The very nature of breaking this logic out into its own function so we can make an explicit choice about invoking it or not, is fulfilling the intent of "make retrieval optional" (or at least is one part of addressing that requirement)

Reply:
@bparees it did not make sense me too. But this was discussed in the backlog refinement meeting and therefore I put it in there. I  will remove that.